File: src/main/scala/fortress/tfol/operations/ClosureEliminator.java
package fortress.tfol.operations;

import fortress.tfol.*;
import fortress.data.NameGenerator;
import java.util.List;
import java.util.Set;
import java.util.HashSet;
import java.util.Map;
import java.lang.IllegalArgumentException;
import fortress.util.Errors;
import java.util.Optional;

// Removes closure given a term, which must be in negation normal form.
// Free variables in the given term are ignored, so the top level term must be
// closed with respect to the signature in question for this operation to be valid.
public class ClosureEliminator {
    private Term toplevelTerm;
    private Map<Type, Integer> scopes;
    private NameGenerator nameGen;
    private Set<FuncDecl> closureFunctions;
    private Set<Term> closureAxioms;
    private ClosureVisitor visitor;
    
    public ClosureEliminator(Term toplevelTerm, Signature signature, Map<Type, Integer> scopes, NameGenerator nameGen) {
        this.toplevelTerm = toplevelTerm;
        this.scopes = scopes;
        this.nameGen = nameGen;
        this.closureFunctions = new HashSet<>();
        this.closureAxioms = new HashSet<>();
        this.visitor = new ClosureVisitor(signature);
    }
    
    public Term convert() {
        return visitor.visit(toplevelTerm);
    }
    
    /**
    * Returns the set of generated closure functions. Must be called after convert.
    */
    public Set<FuncDecl> getClosureFunctions() {
        return closureFunctions;
    }

    /**
    * Returns the set of axioms defining closure functions. Must be called after convert.
    */
    public Set<Term> getClosureAxioms() {
        return closureAxioms;
    }
    
    private class ClosureVisitor extends TermVisitorWithTypeContext<Term> {
        
        public ClosureVisitor(Signature signature) {
            super(signature);
        }
        
        @Override
        public Term visitTop(Top top) {
            return top;
        }
        
        @Override
        public Term visitBottom(Bottom bottom) {
            return bottom;
        }
        
        @Override
        public Term visitVar(Var variable) {
            return variable;
        }
        
        @Override
        public Term visitNot(Not not) {
            return not.mapBody(this::visit);
        }
        
        @Override
        public Term visitAndList(AndList and) {
            return and.mapArguments(this::visit);
        }
        
        @Override
        public Term visitOrList(OrList or) {
            return or.mapArguments(this::visit);
        }
        
        @Override
        public Term visitDistinct(Distinct distinct) {
            throw new IllegalArgumentException("Term supposed to be in NNF but found distinct: " + distinct.toString());
        }
        
        @Override
        public Term visitIff(Iff iff) {
            throw new IllegalArgumentException("Term supposed to be in NNF but found iff: " + iff.toString());
        }
        
        @Override
        public Term visitImplication(Implication imp) {
            throw new IllegalArgumentException("Term supposed to be in NNF but found imp: " + imp.toString());
        }
        
        @Override
        public Term visitEq(Eq eq) {
            return eq.mapArguments(this::visit);
        }
        
        public Term visitApp(App app) {
            return app.mapArguments(this::visit);
        }
        
        public Term visitExistsInner(Exists exists) {
            return Term.mkExists(exists.getVars(), visit(exists.getBody()));
        }

        public Term visitForallInner(Forall forall) {
            return Term.mkForall(forall.getVars(), visit(forall.getBody()));
        }
        
        @Override
        public Term visitDomainElement(DomainElement d) {
            return d;
        }
        
        @Override
        public Term visitTC(TC tc) {
            String relationName = tc.relationName();
            String closureName = "^" + relationName;
            if (!signature.hasFunctionWithName(closureName)) {
                List<Type> argTypes = signature.queryUninterpretedFunctionJava(relationName).get().getArgTypes();
                Type type = argTypes.get(0);
                closureFunctions.add(FuncDecl.mkFuncDecl(closureName, type, type, Type.Bool()));
                Var x = Term.mkVar(nameGen.freshName("x")), y = Term.mkVar(nameGen.freshName("y")), z = Term.mkVar(nameGen.freshName("z"));
                AnnotatedVar ax = x.of(type), ay = y.of(type), az = z.of(type);
                Integer scope = scopes.get(type);
                if (scope < 8) {
                    for (scope /= 2; scope > 1; scope /= 2) {
                        String newFunctionName = nameGen.freshName(relationName);
                        closureFunctions.add(FuncDecl.mkFuncDecl(newFunctionName, type, type, Type.Bool()));
                        closureAxioms.add(Term.mkForall(List.of(ax, ay), Term.mkEq(Term.mkApp(newFunctionName, x, y), Term.mkOr(Term.mkApp(relationName, x, y), Term.mkExists(az, Term.mkAnd(Term.mkApp(relationName, x, z), Term.mkApp(relationName, z, y)))))));
                        relationName = newFunctionName;
                    }
                    closureAxioms.add(Term.mkForall(List.of(ax, ay), Term.mkEq(Term.mkApp(closureName, x, y), Term.mkOr(Term.mkApp(relationName, x, y), Term.mkExists(az, Term.mkAnd(Term.mkApp(relationName, x, z), Term.mkApp(relationName, z, y)))))));
                } else {
                    String helperName = nameGen.freshName(relationName);
                    String reflexiveClosureName = "*" + relationName;
                    closureFunctions.add(FuncDecl.mkFuncDecl(helperName, type, type, type, Type.Bool()));
                    closureFunctions.add(FuncDecl.mkFuncDecl(reflexiveClosureName, type, type, Type.Bool()));
                    Var u = Term.mkVar(nameGen.freshName("u"));
                    closureAxioms.add(Term.mkForall(List.of(ax, ay), Term.mkNot(Term.mkApp(helperName, x, x, y))));
                    closureAxioms.add(Term.mkForall(List.of(ax, ay, az, u.of(type)), Term.mkImp(Term.mkAnd(Term.mkApp(helperName, x, y, u), Term.mkApp(helperName, y, z, u)), Term.mkApp(helperName, x, z, u))));
                    closureAxioms.add(Term.mkForall(List.of(ax, ay, az), Term.mkImp(Term.mkAnd(Term.mkApp(helperName, x, y, y), Term.mkApp(helperName, y, z, z), Term.mkNot(Term.mkEq(x, z))), Term.mkApp(helperName, x, z, z))));
                    closureAxioms.add(Term.mkForall(List.of(ax, ay, az), Term.mkImp(Term.mkAnd(Term.mkApp(helperName, x, y, z), Term.mkNot(Term.mkEq(y, z))), Term.mkApp(helperName, y, z, z))));
                    closureAxioms.add(Term.mkForall(List.of(ax, ay), Term.mkEq(Term.mkApp(reflexiveClosureName, x, y), Term.mkOr(Term.mkApp(helperName, x, y, y), Term.mkEq(x, y)))));
                    Term t = (argTypes.size() == 1) ? Term.mkEq(Term.mkApp(relationName, x), z) : Term.mkApp(relationName, x, z);
                    closureAxioms.add(Term.mkForall(List.of(ax, az), Term.mkImp(Term.mkAnd(t, Term.mkNot(Term.mkEq(x, z))), Term.mkApp(helperName, x, z, z))));
                    closureAxioms.add(Term.mkForall(List.of(ax, ay), Term.mkImp(Term.mkApp(helperName, x, y, y), Term.mkExists(az, Term.mkAnd(t, Term.mkApp(helperName, x, z, y))))));
                    closureAxioms.add(Term.mkForall(List.of(ax, ay), Term.mkEq(Term.mkApp(closureName, x, y), Term.mkExists(az, Term.mkAnd(t, Term.mkApp(reflexiveClosureName, z, y))))));
                }
            }
            return tc.mkApp(closureName).mapArguments(this::visit);
        }
        
        @Override
        public Term visitIntegerLiteral(IntegerLiteral literal) {
            return fortress.util.Errors.<Term>notImplemented();
        }
        
        @Override
        public Term visitBitVectorLiteral(BitVectorLiteral literal) {
            return fortress.util.Errors.<Term>notImplemented();
        }
    }
}

File: src/main/scala/fortress/operations/ClosureEliminator.scala
package fortress.operations

import fortress.msfol._
import fortress.data.NameGenerator
import fortress.util.Errors
import java.lang.IllegalArgumentException
import scala.collection.JavaConverters._

/** Removes closure given a term, which must be in negation normal form.
  * Free variables in the given term are ignored, so the top level term must be
  * closed with respect to the signature in question for this operation to be valid.
  */
class ClosureEliminator(topLevelTerm: Term, signature: Signature, scopes: Map[Sort, Int], nameGen: NameGenerator) {
    val closureFunctions = scala.collection.mutable.Set[FuncDecl]()
    val closureAxioms = scala.collection.mutable.Set[Term]()
    val visitor = new ClosureVisitor
    
    /**
    * Perform the closure elimination and get the resulting term.
    * Don't forget to call getClosureFunctions() and getClosureAxioms()
    * after this.
    * Convert should only be called once per ClosureEliminator object.
    */
    def convert(): Term = {
        visitor.visit(topLevelTerm)
    }
    
    /** Returns the set of generated closure functions. Must be called after convert. */
    def getClosureFunctions: Set[FuncDecl] = closureFunctions.toSet
    
    /** Returns the set of generated skolem functions. Must be called after convert. */
    def getClosureAxioms: Set[Term] =  closureAxioms.toSet
    
    class ClosureVisitor extends TermVisitorWithTypeContext[Term](signature) {
        
        override def visitTop: Term = Top
        
        override def visitBottom: Term = Bottom
        
        override def visitVar(variable: Var): Term = variable
        
        override def visitNot(term: Not): Term = term.mapBody(visit)
        
        override def visitAndList(term: AndList): Term = term.mapArguments(visit)
        
        override def visitOrList(term: OrList): Term = term.mapArguments(visit)
        
        override def visitDistinct(term: Distinct): Term =
            throw new IllegalArgumentException("Term supposed to be in NNF but found distinct: " + term.toString)
        
        override def visitIff(term: Iff): Term =
            throw new IllegalArgumentException("Term supposed to be in NNF but found iff: " + term.toString)
        
        override def visitImplication(term: Implication): Term =
            throw new IllegalArgumentException("Term supposed to be in NNF but found imp: " + term.toString)
        
        override def visitEq(term: Eq): Term = term.mapArguments(visit)
        
        override def visitApp(term: App): Term = term.mapArguments(visit)
        
        override def visitBuiltinApp(term: BuiltinApp): Term = term.mapArguments(visit)
        
        override def visitClosure(c: Closure): Term = {
            var relationName = c.relationName
            val idx = c.arguments.indexOf(c.arg1)
            val closureName = "^" + idx + relationName
            val reflexiveClosureName = "*" + idx + relationName;
            def queryFunction(name: String): Boolean = signature.hasFunctionWithName(name) || closureFunctions.exists(f => f.name == name)
            if (!queryFunction(closureName)) {
                val rel = signature.queryUninterpretedFunction(relationName).get
                var argSorts = rel.getArgSorts
                val sort = argSorts.get(idx)
                closureFunctions += FuncDecl.mkFuncDecl(closureName, argSorts, Sort.Bool)
                val vars = List.tabulate(argSorts.size-2)(_ => Var(nameGen.freshName("bv")))
                val x = Var(nameGen.freshName("x"))
                val y = Var(nameGen.freshName("y"))
                val avars = (List.tabulate(idx)(i => vars(i).of(argSorts.get(i))) :+ x.of(sort) :+ y.of(sort)) ::: List.tabulate(vars.size-idx)(i => vars(idx+i).of(argSorts.get(idx+i+2)))
                val z = Var(nameGen.freshName("z"))
                val az = z.of(sort)
                val scope = scopes(sort)
                def getVarList(v1: Var, v2: Var): List[Var] = (vars.slice(0, idx) :+ v1 :+ v2) ::: vars.slice(idx+2, vars.size)
                if (scope < 8) {
                    for (s <- 1 until scala.math.ceil(scala.math.log(scope)/scala.math.log(2)).toInt) {
                        val newFunctionName = nameGen.freshName(relationName);
                        closureFunctions += FuncDecl.mkFuncDecl(newFunctionName, argSorts, Sort.Bool)
                        closureAxioms += Forall(avars, Iff(App(newFunctionName, getVarList(x, y)), Or(App(relationName, getVarList(x, y)), Exists(az, And(App(relationName, getVarList(x, z)), App(relationName, getVarList(z, y)))))))
                        relationName = newFunctionName;
                    }
                    closureAxioms += Forall(avars, Iff(App(closureName, getVarList(x, y)), Or(App(relationName, getVarList(x, y)), Exists(az, And(App(relationName, getVarList(x, z)), App(relationName, getVarList(z, y)))))))
                } else if (queryFunction(reflexiveClosureName)) {
                    closureAxioms += Forall(avars, Iff(App(closureName, getVarList(x, y)), Exists(az, And(App(relationName, getVarList(x, z)), App(reflexiveClosureName, getVarList(z, y))))))
                } else {
                    val helperName = nameGen.freshName(relationName);
                    closureFunctions += FuncDecl.mkFuncDecl(reflexiveClosureName, argSorts, Sort.Bool);
                    argSorts.add(sort)
                    closureFunctions += FuncDecl.mkFuncDecl(helperName, argSorts, Sort.Bool);
                    val u = Var(nameGen.freshName("u"));
                    closureAxioms += Forall(avars, Not(App(helperName, getVarList(x, x) :+ y)))
                    closureAxioms += Forall(avars :+ az :+ u.of(sort), Implication(And(App(helperName, getVarList(x, y) :+ u), App(helperName, getVarList(y, z) :+ u)), App(helperName, getVarList(x, z) :+ u)))
                    closureAxioms += Forall(avars :+ az, Implication(And(App(helperName, getVarList(x, y) :+ y), App(helperName, getVarList(y, z) :+ z), Not(Eq(x, z))), App(helperName, getVarList(x, z) :+ z)))
                    closureAxioms += Forall(avars :+ az, Implication(And(App(helperName, getVarList(x, y) :+ z), Not(Eq(y, z))), App(helperName, getVarList(y, z) :+ z)))
                    closureAxioms += Forall(avars, Implication(And(App(relationName, getVarList(x, y)), Not(Eq(x, y))), App(helperName, getVarList(x, y) :+ y)))
                    closureAxioms += Forall(avars, Implication(App(helperName, getVarList(x, y) :+ y), Exists(az, And(App(relationName, getVarList(x, z)), App(helperName, getVarList(x, z) :+ y)))))
                    closureAxioms += Forall(avars, Iff(App(reflexiveClosureName, getVarList(x, y)), Or(App(helperName, getVarList(x, y) :+ y), Eq(x, y))))
                    closureAxioms += Forall(avars, Iff(App(closureName, getVarList(x, y)), Exists(az, And(App(relationName, getVarList(x, z)), App(reflexiveClosureName, getVarList(z, y))))))
                }
            }
            App(closureName, c.arguments).mapArguments(visit)
        }
        
        override def visitReflexiveClosure(rc: ReflexiveClosure): Term = {
            var relationName = rc.relationName
            val idx = rc.arguments.indexOf(rc.arg1)
            val reflexiveClosureName = "*" + idx + relationName
            val closureName = "^" + idx + relationName;
            def queryFunction(name: String): Boolean = signature.hasFunctionWithName(name) || closureFunctions.exists(f => f.name == name)
            if (!queryFunction(reflexiveClosureName)) {
                val rel = signature.queryUninterpretedFunction(relationName).get
                var argSorts = rel.getArgSorts
                val sort = argSorts.get(idx)
                closureFunctions += FuncDecl.mkFuncDecl(reflexiveClosureName, argSorts, Sort.Bool)
                val vars = List.tabulate(argSorts.size-2)(_ => Var(nameGen.freshName("bv")))
                val x = Var(nameGen.freshName("x"))
                val y = Var(nameGen.freshName("y"))
                val avars = (List.tabulate(idx)(i => vars(i).of(argSorts.get(i))) :+ x.of(sort) :+ y.of(sort)) ::: List.tabulate(vars.size-idx)(i => vars(idx+i).of(argSorts.get(idx+i+2))) 
                val z = Var(nameGen.freshName("z"))
                val az = z.of(sort)
                val scope = scopes(sort)
                def getVarList(v1: Var, v2: Var): List[Var] = (vars.slice(0, idx) :+ v1 :+ v2) ::: vars.slice(idx+2, vars.size)
                if (scope > 8) {
                    val helperName = nameGen.freshName(relationName);
                    argSorts.add(sort)
                    closureFunctions += FuncDecl.mkFuncDecl(helperName, argSorts, Sort.Bool);
                    val u = Var(nameGen.freshName("u"));
                    closureAxioms += Forall(avars, Not(App(helperName, getVarList(x, x) :+ y)))
                    closureAxioms += Forall(avars :+ az :+ u.of(sort), Implication(And(App(helperName, getVarList(x, y) :+ u), App(helperName, getVarList(y, z) :+ u)), App(helperName, getVarList(x, z) :+ u)))
                    closureAxioms += Forall(avars :+ az, Implication(And(App(helperName, getVarList(x, y) :+ y), App(helperName, getVarList(y, z) :+ z), Not(Eq(x, z))), App(helperName, getVarList(x, z) :+ z)))
                    closureAxioms += Forall(avars :+ az, Implication(And(App(helperName, getVarList(x, y) :+ z), Not(Eq(y, z))), App(helperName, getVarList(y, z) :+ z)))
                    closureAxioms += Forall(avars, Implication(And(App(relationName, getVarList(x, y)), Not(Eq(x, y))), App(helperName, getVarList(x, y) :+ y)))
                    closureAxioms += Forall(avars, Implication(App(helperName, getVarList(x, y) :+ y), Exists(az, And(App(relationName, getVarList(x, z)), App(helperName, getVarList(x, z) :+ y)))))
                    closureAxioms += Forall(avars, Iff(App(reflexiveClosureName, getVarList(x, y)), Or(App(helperName, getVarList(x, y) :+ y), Eq(x, y))))
                } else if (queryFunction(closureName)) {
                    closureAxioms += Forall(avars, Iff(App(reflexiveClosureName, getVarList(x, y)), Or(Eq(x, y), App(closureName, getVarList(x, y)))))
                } else {
                    closureFunctions += FuncDecl.mkFuncDecl(closureName, argSorts, Sort.Bool)
                    for (s <- 1 until scala.math.ceil(scala.math.log(scope)/scala.math.log(2)).toInt) {
                        val newFunctionName = nameGen.freshName(relationName);
                        closureFunctions += FuncDecl.mkFuncDecl(newFunctionName, argSorts, Sort.Bool)
                        closureAxioms += Forall(avars, Iff(App(newFunctionName, getVarList(x, y)), Or(App(relationName, getVarList(x, y)), Exists(az, And(App(relationName, getVarList(x, z)), App(relationName, getVarList(z, y)))))))
                        relationName = newFunctionName;
                    }
                    closureAxioms += Forall(avars, Iff(App(closureName, getVarList(x, y)), Or(App(relationName, getVarList(x, y)), Exists(az, And(App(relationName, getVarList(x, z)), App(relationName, getVarList(z, y)))))))
                    closureAxioms += Forall(avars, Iff(App(reflexiveClosureName, getVarList(x, y)), Or(Eq(x, y), App(closureName, getVarList(x, y)))))
                }
            }
            App(reflexiveClosureName, rc.arguments).mapArguments(visit)
        }
        
        override def visitForallInner(term: Forall): Term = term.mapBody(visit)
        
        override def visitExistsInner(term: Exists): Term = term.mapBody(visit)
        
        override def visitDomainElement(d: DomainElement): Term = d
        
        override def visitIntegerLiteral(literal: IntegerLiteral): Term = literal
        
        override def visitBitVectorLiteral(literal: BitVectorLiteral): Term = literal
        
        override def visitEnumValue(e: EnumValue): Term = e
        
    }
}